Add and update security designs for Contexts (#274)
ros2/design: articles/ros2_security_contexts.md

---
layout: default
title: ROS 2 Security Contexts
permalink: articles/ros2_security_contexts.html
abstract:
  This article specifies the integration between security and contexts.
author:  >
  [Ruffin White](https://github.com/ruffsl),
  [Mikael Arguedas](https://github.com/mikaelarguedas)

published: false
categories: Security
---

{:toc}

# {{ page.title }}

<div class="abstract" markdown="1">
{{ page.abstract }}
</div>

Original Author: {{ page.author }}

TODO: Some concise overview introduction here

## Concepts

Before detailing the SROS 2 integration of the contexts, the following concepts are introduced.

### Participant

> Participant is the object representing a single entity on the network, in the case of DDS the ``Participant`` is a DDS DomainParticipant and the object to which a set of access control and security identity apply.

Reviewer: This could be phrased a little clearer.

```suggestion
Participant is the object representing a single entity on the network. In the case of DDS, the ``Participant`` is a DDS DomainParticipant, which has both access control permissions and a security identity.
```

### Namespaces

Namespaces are a fundamental design pattern in ROS and are widely used to organize and differentiate many types of resources as to be uniquely identifiable; i.e. for topics, services, actions, and node names.

> As such, the concept of namespacing is well know and understood by current users, as well as strongly supported with the existing tooling.

Reviewer: ```suggestion
As such, the concept of namespacing is well known and understood by current users, as well as strongly supported by the existing tooling.
```

> As such, the concept of namespaceing is well know and understood by current users, as well as strongly supported with the existing tooling.

Reviewer 2: namespaceing -> namespacing

Author: ```suggestion
As such, the concept of namespacing is well know and understood by current users, as well as strongly supported with the existing tooling.
```

Namespaces are often configurable at runtime via command line arguments or statically/programmatically via launch file declarations.

Previously, the Fully Qualified Name (FQN) of a node was used directly by a selected security directory lookup strategy to load the necessary key material.

> However, with the advent of contexts, such a direct mapping of FQN to security artifacts may no longer suffice.

Reviewer 2: Contexts were already there (see for example https://github.com/ros2/rclcpp/blob/23719237618ad15f3d074af316cdeabc077f0094/rclcpp/include/rclcpp/context.hpp#L53).
The difference is that now a `Participant` will map to a `Context` and no more to a `Node`.

Author: Is there another design doc for this change we could just defer the reader to?

Reviewer 2: > Is there another design doc for this change we could just defer the reader to?

mmm, I don't think so.

The original motivation of `Context` was an object to encapsulate all "global" data used in an init/shutdown cycle.

Reviewer 3: Agreed. What we try to convey here is that before contexts were transparent to security and now the new mapping of participant to contexts make the use of node FQN not adapted for security artifacts mapping anymore.

How about: "However, now that ``Participants`` map to contexts and not nodes, such a direct mapping of node FQN to security artifacts is no longer appropriate."

Reviewer: Agreed. @mikaelarguedas's rephrase seems perfect to me.

Author: ```suggestion
However, now that Participants map to contexts and not nodes, such a direct mapping of node FQN to security artifacts is no longer appropriate.
```

### Contexts

With the advent of ROS 2, multiple nodes may now be composed into one process for improved performance.
Previously however, each node would retain it's one to one mapping to a separate middleware ``Participant``.
Given the non-negligible overhead incurred of multiple ``Participant``s per process, a concept of contexts was introduced.

> Contexts permit a many-to-one mapping of nodes to ``Participant`` by grouping many nodes per context, and one ``Participant`` per context.

Reviewer: This will need a slight rephrase as well, to reflect that contexts already existed.

```suggestion
Given the non-negligible overhead incurred of multiple ``Participant``s per process, a change was introduced to map a single ``Participant`` to a context, and allow for multiple nodes to share that context.
```

Based on the DDS Security specification v1.1, a ``Participant`` can only utilise a single security identity; consequently the access control permissions applicable to every node mapped to a given context must be consolidated and combined into a single set of security artifacts.
As such, additional tooling and extensions to SROS 2 are necessary to support this new paradigm.

## Keystore

With the additional structure of contexts, it’s perhaps best to take the opportunity to restructure the keystore layout as well.
Rather than a flat directory of namespaced node security directories, we can push all such security directories down into a designated `contexts` sub-folder.
Similarly, private and public keystore materials can also be pushed down into their own respective sub-folders within the root keystore directory.
This is reminiscent of the pattern used earlier Keymint [1].

```
$ tree keystore/
keystore
├── contexts
│   └── ...
│       └── ...
├── private
│   ├── ca.csr.pem
│   └── ca.key.pem
└── public
    ├── ca.cert.pem
    ├── identity_ca.cert.pem
    └── permissions_ca.cert.pem
```

### ``public``

> The ``public`` directory contains anything permissable as public, such as public certificates for the identity or permissions certificate authorities.

Reviewer 2: permissable -> permissible

Author: ```suggestion
The ``public`` directory contains anything permissible as public, such as public certificates for the identity or permissions certificate authorities.
```

As such, this can be given read access to all executables.
Note that in the default case, both the `identity_ca` and `permissions_ca` are the same CA certificate.

### ``private``

The ``private`` directory contains anything permissable as private, such as private key material for aforementioned certificate authorities.

> This directory should be redacted before deploying the keystore onto the target device/robot.

Reviewer 2: What do you mean by "redacted" in this context? populated? (non-native speaker here)

Author: redacted, removed, deleted. Whatever is clearer. The private folder in the keystore just shouldn't be copied to the eventual target to be used at run-time; instead, store these private keys of the CA in safe location.

Reviewer 2: Ah ok, now I understand.

### ``contexts``

The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.

## Runtime

TODO: Some transition paragraph here about ros launch

### Unqualified context path

For nodes with unqualified context paths, the context directory will subsequently default to the root level context.

``` xml
<launch>
  <node pkg="demo_nodes_cpp" exec="talker"/>
  <node pkg="demo_nodes_cpp" exec="listener"/>
</launch>
```

```
$ tree contexts/
contexts/
├── cert.pem
├── governance.p7s
├── identity_ca.cert.pem
├── key.pem
├── permissions_ca.cert.pem
└── permissions.p7s
```

### Pushed unqualified context path

For nodes with unqualified context paths pushed by a namespace, the context directory will subsequently be pushed to the relative sub-folder.

``` xml
<launch>
  <node pkg="demo_nodes_cpp" exec="talker"/>
  <group>
    <push_ros_namespace namespace="foo"/>
    <node pkg="demo_nodes_cpp" exec="listener"/>
  </group>
</launch>

> ```

Reviewer 2: Should this other example work differently?
``` xml
<launch>
  <node pkg="demo_nodes_cpp" exec="talker"/>
  <node pkg="demo_nodes_cpp" exec="listener" namespace="foo"/>
</launch>
```

I imagine we should add new arguments to the node action like: `context_name` and `context_namespace`.
In the example using `push_ros_namespace`, the context namespace is also affected and the security directory is pushed to the relative sub-folder.
In the second example using the `namespace` attribute, `context_namespace` is not modified and the root security directory is used.

Does that sound correct?

Author: I hadn't thought about the order of operations that would play out if the user specified both the context and namespace XML attributes, but I was banking on just relying on however RCL would normally resolve the namespace arrangement, and a unqualified context name would just be appended to the resulting namespace.

> I imagine we should add new arguments to the node action like: context_name and context_namespace.
In the example using push_ros_namespace, the context namespace is also affected and the security directory is pushed to the relative sub-folder.
In the second example using the namespace attribute, context_namespace is not modified and the root security directory is used.

I think that would feel inconsistent to the user:  given the namespace pushing outside the node would effect the context, while pushing the namespace in the node element would not. Having two context attributes to divide the context path I think would also escalate boilerplate when seeking to set the context via args.

Reviewer 2: > I think that would feel inconsistent to the user: given the namespace pushing outside the node would effect the context, while pushing the namespace in the node element would not

Yes, I agree it's confusing and probably not useful.

> Having two context attributes to divide the context path I think would also escalate boilerplate when seeking to set the context via args.

:+1:  Yes, I think that having just one attribute for the context name is reasonable.

Author: Resolving of default or relative context path using roslaunch is clarified in https://github.com/ros2/design/pull/274/files#diff-0c022ad3d42831e4e8cf2336da70d354R104

```
$ tree --dirsfirst contexts/
contexts/
├── foo
│   ├── cert.pem
│   ├── governance.p7s
│   ├── identity_ca.cert.pem
│   ├── key.pem
│   ├── permissions_ca.cert.pem
│   └── permissions.p7s
├── cert.pem
├── governance.p7s
├── identity_ca.cert.pem
├── key.pem
├── permissions_ca.cert.pem
└── permissions.p7s
```

### Relatively pushed qualified context path

For nodes with qualified context paths pushed by a namespace, the qualified context directory will subsequently be pushed to the relative sub-folder.

``` xml
<launch>
  <group>
    <push_ros_namespace namespace="foo"/>

>     <node pkg="demo_nodes_cpp" exec="listener" context="bar"/>

Reviewer 2: I would use `context_name` instead of just `context`.

Author: I'm not sure what to call it, as it could be a _path or just a relative _name, so I left it as context.

Reviewer 2: If we left only one attribute for specifying the context name and not two, using just `context` is fine.

There is a mismatch between the middleware credentials/permissions loaded and the resources accessible within the process.

By using contexts, all nodes in a context share the same security identity and access control credentials.
This inevitably means that code compiled to node ``foo`` can access credentials/permissions only trusted to node ``bar``.
This consequence of composition could unintendedly subvert the minimal spanning policy as architected by the policy designer or measured/generated via ROS 2 tooling/IDL.

With the introduction of contexts, it becomes possible to describe the union of access control permission by defining a collection of SROS 2 policy profiles as element within a specific context.
This would allow for formal analysis tooling [2] to check for potential violations in information flow control given the composing of nodes at runtime.
If a process contains a single context, this reconciles the permissions of a ``Participant`` and the ones of the process.

However, should multiple contexts be used per process, then such security guaranties are again lost because both contexts will share the same memory space.

> Thus it should be asked whether if multiple contexts per process should even be supported.

Reviewer 2: I think that the main reason to map a `Participant` to a `Context` and not one `Participant` per process, is to allow domain bridge applications.

Author: For such bridge applications, what would be a layer of abstraction to model the union of contexts over different domains? I'd like to empathize this question, as bridge interfaces are particularly targeted for exploiting flaws in access control and information flow control. Think of CAN bus to infotainment systems, or LAN to AP radios. Thus retaining a level of abstraction to encapsulate this, be it contexts or something else, could go a long way in aiding formal analyses of such precarious interfaces.

Reviewer 2: IIUC, to bridge two `Domains` in DDS you need two `DomainParticipants` (one in each domain). Each `DomainParticipant` needs their own security artifacts.
I don't understand why another level of abstraction would be needed to encapsulate this.

Author: It's not so much about generating secure DDS artifacts, and more about modeling the permissions that process may have as a whole. Let's say you have two Bridge processes that interface with domains A and C respectively, and communicate with each other over domain B. If I had a particularly sensitive topic in a, how would a design tool be used to check/verify that it's propagation to domain C is impossible. If this was over one domain, the current model of context could be used to associate the subscribe access of the first Bridge node as a potential information leak to subscribers of the second bridge note, given the transitive probation of data from the first bridge flowing into the second bridge.

Author: Compromise reached by passing necessary metadata for bridging in policy context.
See e387ac17364ef89e90ff7a9e68d475350347b4bf

In summary, the distinction here is that before, the composition of multiple permissions could not be conveyed to the tooling.
Whether nodes could gain the permission of others in the same process space is not the hinge point of note; it's the fact that such side effects could not be formally modeled or accounted for by the designer.
It will now be possible with contexts, however allowing for multiple contexts per process will reintroduce and exacerbates the same modeling inaccuracies.

### Composable launchfile includes

A particular challenge in using launchfiles with security contexts is that of keeping the include hierarchy composable.
An inherit tradeoff between simplicity and configurability can arise when writing launchfiles for downstream use.
Authors can selectively choose what attributes to expose as input arguments, while users may implicitly override provided defaults.

In case of contexts, it is not inherently clear what best practices either package authors or users should employ to retain a composable and intuitive launchfile structure. E.g:

> Should authors parametrize context paths for each node as input arguments?

Reviewer 2: IMO, yes

Author: @mikaelarguedas and I where wondering if that would make things too noisy or add to the boilerplate in launchfiles. Given node name uniqueness, we could encourage users to just name there relative context the same as the node name by convention, and rely on pushing namespaces. (I guess that gets us back to why not just default the relative context name being the node name for single process nodes)

The only deviation would be when composing nodes into containers. Is there an example we can look at on how one may create a container in a parent launchfile, then include a child launchfile by passing the container (reference/name/or-what?) its' nodes are expected to compose into.

Reviewer 2: > I guess that gets us back to why not just default the relative context name being the node name for single process nodes.

I think that that is a good idea. See my other comment: https://github.com/ros2/design/pull/274#issuecomment-592071949.

Reviewer 2: > The only deviation would be when composing nodes into containers. Is there an example we can look at on how one may create a container in a parent launchfile, then include a child launchfile by passing the container (reference/name/or-what?) its' nodes are expected to compose into.

I can think there are a few ways of achieving that, but I don't think we have a recommended way of doing it.

Author: Statement formalized here https://github.com/ros2/design/pull/274/files#diff-0c022ad3d42831e4e8cf2336da70d354R240

> Should users push namespaces of included launchfiles to separate contexts?

Reviewer 2: If the included launch file has an argument to modify the context name/namespace, it wouldn't be necessary.

Author: Lets merge this line of thought back to this thread: https://github.com/ros2/design/pull/274#discussion_r384948942

> Should the setting of security environment variables be discouraged from within launchfiles, limiting the use of simple static analysis of launchfiles combined with Node IDL for procedural context generation?

Reviewer 2: IMO, yes.

Author: Yeah, I was thinking of other patterns that could be used to specify or override the context and wanted to point out a potential anti-pattern.

Author: This is now formal stated here https://github.com/ros2/design/pull/274/files#diff-0c022ad3d42831e4e8cf2336da70d354R274

### Composable nodes in container

Given that containers can be dynamic, where nodes can be added or removed at runtime, there is perhaps some question as to how containers should integrate with secure contexts.
The most straightforward approach is perhaps only specifying the context wherever the container is first defined/launched, applying to the container participant, thus to all nodes/components inside that container.

> This should be further deliberated when eventually extending the launch API for containers.

Reviewer 2: The components container inherit from the `Node` action, so the context name and namespace could also be configured in the same way as in `Node`.
`ComposableNode` descriptions shouldn't change.

Author: Nice, that's simple. So `ComposableNode` would just not have a context XML attribute?

Reviewer 2: > Nice, that's simple. So ComposableNode would just not have a context XML attribute?

Exactly, it wouldn't be possible to have one as those components will be loaded in the same `Context` (a node container only has one `Context` where all the nodes are loaded).

Author: Use of one context per container also stated here https://github.com/ros2/design/pull/274/files#diff-0c022ad3d42831e4e8cf2336da70d354R104

The `ROS_SECURITY_ROOT_DIRECTORY` environment variable should by convention point to this directory.

## Runtime

Given the normative case where a context within a policy may be specific to a single node/container process, the namespace the node is remapped to will inevitably affect the required security permissions necessary within the context.
To highlight this interdependency, and to help avoid context path collisions, a hierarchy borrowing namespaces is appropriate. 
By convention, roslaunch could be used to prefix relative context paths for single process node or containers using the namespace in scope, to enable a convention of composable launch files with adjustable and parameterized context paths.
Given the runtime command argument for specifying the fully qualified context path, roslaunch would accordly resolve relative context paths for executables, as defined by launch attributes.

### Unqualified context path

For single process nodes with unqualified context paths, the context directory will subsequently default to the root level context.

``` xml
<launch>
  <node pkg="demo_nodes_cpp" exec="talker"/>
  <node pkg="demo_nodes_cpp" exec="listener"/>
</launch>

> ```

Reviewer 2: Considering the previous discussion, we can't do any assumption about how the context name/path relates to the node name. I would delete the whole discussion about launch files, and just put an example of how explicitly pass a context name:

``` xml
<launch>
  <node pkg="demo_nodes_cpp" exec="talker" context="talker"/>
  <node pkg="demo_nodes_cpp" exec="listener"  context="listener"/>
</launch>
```

If the context name/path isn't specified, we can take `"/"` or `""` as the default.

P.S.: There's a lot of improvement to do in ROS2 about launchfile reuse, e.g.: there's no recommended pattern to add components to a container, etc.

Author: Ah, I was still thinking, regardless of RCL only being able to take fully qualified context paths, ros launch could still be used to resolve relative paths to fully qualified paths before reaching RCL by using the namespace in scope (regardless of the namespace code may hardcode to), as just supporting convention of composing launch files. We can narrow this down at the security WG meeting today.

---
layout: default
title: ROS 2 Security Contexts
permalink: articles/ros2_security_contexts.html
abstract:
  This article specifies the integration between security and contexts.
author:  >
  [Ruffin White](https://github.com/ruffsl),
  [Mikael Arguedas](https://github.com/mikaelarguedas)

> published: false

Reviewer 2: I would change this to `true`.

Author: Done in f75310f

This is reminiscent of the pattern used earlier Keymint [1].

```
$ tree keystore/
keystore
├── contexts
│   └── ...
│       └── ...
├── private
│   ├── ca.csr.pem
│   └── ca.key.pem
└── public
    ├── ca.cert.pem
    ├── identity_ca.cert.pem -> ca.cert.pem
    └── permissions_ca.cert.pem -> ca.cert.pem
```

### ``public``

The ``public`` directory contains anything permissible as public, such as public certificates for the identity or permissions certificate authorities.
As such, this can be given read access to all executables.
Note that in the default case, both the `identity_ca` and `permissions_ca` points to the same CA certificate.

### ``private``

The ``private`` directory contains anything permissable as private, such as private key material for aforementioned certificate authorities.
This directory should be redacted before deploying the keystore onto the target device/robot.

### ``contexts``

The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.

> The `ROS_SECURITY_ROOT_DIRECTORY` environment variable should by convention point to this directory.

Reviewer 3: there is something that doesnt feel right about this.
as this is the "ROOT_DIRECTORY" it seems like I should be able to just ship that to my robot and it would "just work".
But files in the subfolders of `contexts` symlink to files outside of the "ROOT_DIRECTORY".

Also given that all the sros2 commands use the keystore directory as argument, it seems weird to say that the "root" is actually a subfolder of that.

I don't really remember, what was the rationale for not keeping `ROS_SECURITY_ROOT_DIRECTORY` pointing to the keystore directory ? to not append "contexts" arbitrarily in RCL ?
If it still makes sense to keep if pointing to `contexts`, could we change the env var name to something without "root" in it ? like `ROS_SECURITY_CONTEXTS_DIRECTORY` or something to that effect?

Author: > But files in the subfolders of contexts symlink to files outside of the "ROOT_DIRECTORY".
> Also given that all the sros2 commands use the keystore directory as argument, it seems weird to say that the "root" is actually a subfolder of that.

That's why I suggested that RCL should append the `contexts` sub folder too the root when knowing to look for context security artifacts, so that the root dir would span over any needed `public`, `contexts`, etc, directories. Given we are already hard coding conventions for artifact file names, it doesn't seem much of a stretch to hardcode conventions in the root directory structure.

> If it still makes sense to keep if pointing to contexts, could we change the env var name to something without "root" in it ? like ROS_SECURITY_CONTEXTS_DIRECTORY or something to that effect?

That could work, but I think reverting/correcting the ROOT env to point to same root keystore dir the the sros2 cli expects would help maintain consistency in user experience, as well as future proof the use of that ENV when we allow for setting more advance URI strings, other than just file system paths, like if using a TPM or remote keybase.

I left such comments on https://github.com/ros2/rcl/pull/515 , but it looks like github swallowed my remarks when the PR was rebased (losing review comments like that is annoying). :<

Reviewer 2: > I left such comments on ros2/rcl#515 , but it looks like github swallowed my remarks when the PR was rebased (losing review comments like that is annoying). :<

Sorry if I lost a comment, but I needed to rebase.
Pointing to the `keystore` and not to the `contexts` directory sounds good to me.

I think that both `rcl` and `test_security` need to be updated.

Let me know if you're planning to send a PR, if not I will try to do it on Monday.

Author: > Sorry if I lost a comment, but I needed to rebase.

Don't worry, not faulting you, I blame github.
Just wasn't sure why comments are preserved differently than reviews.

See 09e5c9e for what I was first envisioning, and let me know if you see an issue.

Reviewer 2: > See 09e5c9e for what I was first envisioning, and let me know if you see an issue.

SGTM!

Reviewer 2: Fixed in https://github.com/ros2/rcl/pull/607.

## Runtime

Given the normative case where a context within a policy may be specific to a single node/container process, the namespace the node is remapped to will inevitably affect the required security permissions necessary within the context.
To highlight this interdependency, and to help avoid context path collisions, a hierarchy borrowing namespaces is appropriate. 

> By convention, ros2launch could be used to prefix relative context paths for single process node or containers using the namespace in scope, to enable a convention of composable launch files with adjustable and parameterized context paths.

Reviewer 3: Is this still something we consider ?
I thought we ended up moving away from any correlation between node name or namespace and context paths

Author: We moved away from node names. But the namespace in scope of the launch file could still be used to resolve relative context paths.

Reviewer 2: I think that the same applies to namespaces.
If a context have several nodes, each of them created with a different `namespaces`, which of them will be used to expand the `security context`?

Author: > I think that the same applies to namespaces.

I think that was already addressed in that the context would be resolved at the point where the composing container is first instantiated in the launch file, e.g. what ever relative or absolute context the user prescribes when starting the container that may house nodes of various other namespaces. 

> > Nice, that's simple. So ComposableNode would just not have a context XML attribute?
>
> Exactly, it wouldn't be possible to have one as those components will be loaded in the same Context (a node container only has one Context where all the nodes are loaded).
https://github.com/ros2/design/pull/274/files#r385215843

From what you also mentioned prior, roslaunch does not yet have an API for instantiating containers, so that would also need to be addressed:

https://github.com/ros2/design/pull/274#discussion_r390290150

Reviewer 3: Outside of the case of containers. I may want my nodes (standalone processes) to be pushed in a namespace without wanting them to look for security files in that same namespace. for example:
I want to have all my navigation nodes under the "navigation" namespace:
  
upstream launch file : 
awesome_navigation.launch.xml
```xml
<launch>
  <node ... awesome navigation node />
  <node static transform publisher context="tf_publisher" />
</launch>
```
my launchfile
```xml
<launch>
<group>
    <push_ros_namespace namespace="navigation"/>
    <include pkg="awesome_navigation" file="awesome_navigation.launch.xml"/>
<group>
</launch>
```

I can't use my generic security context "tf_publisher" because now it will be looking for "/foo/tf-publisher".
All I can do is:
- go modify the upstream launch file to use an absolute context
or
- create a `/foo/tf_publisher` set of artifacts to be used just by this node.

---

The story around composing nodes and launchfiles with security contexts is not clear, and neither the suggested approach or following namespaces or the alternatives allows to address it. It seems that the only discussed alternative allowing to modify included launchfiles contexts in a granular manner is to have to add a bunch of arguments to each launchfile

Author: > I may want my nodes (standalone processes) to be pushed in a namespace without wanting them to look for security files in that same namespace.

Some concrete reasons for doing so might help us polish our design ideas. 

> I can't use my generic security context "tf_publisher" because now it will be looking for "/foo/tf-publisher".

To be fair, as a user, you did choose to push the entire 3rd party launch file into a custom namespace, so I think it makes sense that the default behavior for all such context instances pushed would be to similarly push the relative context path they should look in.

> All I can do is

There is another alternative: to customize your keystore contexts structure to match the level of complexity that you'd like to achieve for the context lookup.
On could simply use links in the keystore:

```
cd ~/keystore/contexts/foo
ln -s ../tf_publisher tf_publisher
```

> It seems that the only discussed alternative allowing to modify included launchfiles contexts in a granular manner is to have to add a bunch of arguments to each launchfile

Which upstream package authors could choose or not choose to follow through with, that being I think most would skip on given how tediously verbose it quickly gets. (Security & Usability)

> I can't use my generic security context "tf_publisher" because now it will be looking for "/foo/tf-publisher".

I don't think it would be a good habit to have external launch files starting processes that use contexts orthogonal to namespace hierarchy, as it might get harder for the user to grok what part of the computational graph uses what security certificates.

Also given the static transform publisher would be remapped by the namespace scope (unless you pass a ns="/" arg that I didn't see), the node will need unique permissions like node param services that are FQN'ed.

In summery, if the launch file is in your control, then using absolute context paths could be a simple means of overriding name space scoping when complexity in context lookup is required. Alternatively, when the launch file is upstream, then you can still compensate by adjusting the keystore hierarchy which is still in your control.

Author: Relegated to future work: 3e148b2

Given the runtime command argument for specifying the fully qualified context path, ros2launch would accordingly resolve relative context paths for executables, as defined by launch attributes.

### Unqualified context path

For single process nodes with unqualified context paths, the context directory will subsequently default to the root level context.

``` xml
<launch>
  <node pkg="demo_nodes_cpp" exec="talker"/>
  <node pkg="demo_nodes_cpp" exec="listener"/>
</launch>
```

```
$ tree contexts/
contexts/
├── cert.pem
├── governance.p7s
├── identity_ca.cert.pem -> ../public/identity_ca.cert.pem
├── key.pem
├── permissions_ca.cert.pem -> ../public/permissions_ca.cert.pem
└── permissions.p7s
```

### Pushed unqualified context path

> For single process nodes with unqualified context paths pushed by a namespace, the context directory will subsequently be pushed to the relative sub-folder.

Reviewer 3: Same here, is this still something we consider?

Author: https://github.com/ros2/design/pull/274#discussion_r403319180

Author: Relegated to future work: 3e148b2

Original Author: {{ page.author }}

> This design document formalizes the integration of contexts in ROS2 with security.

Reviewer 3: ```suggestion
This design document formalizes the integration of contexts in ROS 2 with security.
```

> In summary, all context's within a process must load a common context path that contains the unique runtime security artifacts.

Reviewer 3: ```suggestion
In summary, all contexts within a process must load a common context path that contains the unique runtime security artifacts.
```

Reviewer 3: What do we mean by unique here ?

Author: That the security artifacts are unique to that context path, not necessarily unique to a process or rcl context instance. Have a suggested rephrasing?

Author: Rephrased in 7473326b3658e647bd048f742b77a49bf825d5f3

The ``contexts`` directory contains the security artifacts associated with individual contexts, and thus node directories are no longer relevant.
Similar to node directories however, the `contexts` folder may still recursively nest sub-paths for organizing separate contexts.
The `ROS_SECURITY_ROOT_DIRECTORY` environment variable should by convention point to this directory.

## Runtime

Given the normative case where a context within a policy may be specific to a single node/container process, the namespace the node is remapped to will inevitably affect the required security permissions necessary within the context.
To highlight this interdependency, and to help avoid context path collisions, a hierarchy borrowing namespaces is appropriate. 

> By convention, roslaunch could be used to prefix relative context paths for single process node or containers using the namespace in scope, to enable a convention of composable launch files with adjustable and parameterized context paths.

Reviewer 3: ```suggestion
By convention, ros2launch could be used to prefix relative context paths for single process node or containers using the namespace in scope, to enable a convention of composable launch files with adjustable and parameterized context paths.
```

> Given the runtime command argument for specifying the fully qualified context path, roslaunch would accordly resolve relative context paths for executables, as defined by launch attributes.

Reviewer 3: ```suggestion
Given the runtime command argument for specifying the fully qualified context path, ros2launch would accordingly resolve relative context paths for executables, as defined by launch attributes.
```

In case of contexts, it is not inherently clear what best practices either package authors or users should employ to retain a composable and intuitive launchfile structure. E.g:
Should authors parametrize context paths for each node as input arguments?
Should users push namespaces of included launchfiles to separate contexts?

To be sure though, the setting of security environment variables from within launchfiles should be discouraged, as this would restrict the use of static analysis of launchfiles combined with Node IDL for procedural context generation.

### Composable nodes in container

Given that containers can be dynamic, where nodes can be added or removed at runtime, there is perhaps some question as to how containers should integrate with secure contexts.

> In roslaunch, the namespace in scope at the container's instantiation could be used to resolve the container's specified relative context path, thus to all nodes/components inside that container.

Reviewer 3: ```suggestion
In ros2launch, the namespace in scope at the container's instantiation could be used to resolve the container's specified relative context path, thus to all nodes/components inside that container.
```
